dotnet/src/Connectors/Connectors.Memory.Kusto/KustoMemoryStore.cs

```diff
@@ -0,0 +1,411 @@
+                CslCommandGenerator.GenerateTablesShowCommand() + $" | where TableName startswith '{c_collectionPrefix}'",
+                GetClientRequestProperties()
+            ).ConfigureAwait(false);
+
+        foreach (var item in result)
+        {
+            yield return GetCollectionName(item.TableName);
+        }
+    }
+
+    /// <inheritdoc/>
+    public async Task<(MemoryRecord, double)?> GetNearestMatchAsync(string collectionName, Embedding<float> embedding, double minRelevanceScore = 0, bool withEmbedding = false, CancellationToken cancellationToken = default)
+    {
+        var result = this.GetNearestMatchesAsync(collectionName, embedding, 1, minRelevanceScore, withEmbedding, cancellationToken);
+        return await result.FirstOrDefaultAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(string collectionName, Embedding<float> embedding, int limit, double minRelevanceScore = 0, bool withEmbeddings = false, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var similarityQuery = $"{this.GetBaseQuery(collectionName)} | extend similarity=series_cosine_similarity_fl('{KustoSerializer.SerializeEmbedding(embedding)}', {s_collectionColumns[2].Name}, 1, 1)";
+
+        if (minRelevanceScore != 0)
+        {
+            similarityQuery += $" | where similarity > {minRelevanceScore}";
+        }
+
+        similarityQuery += $" | top {limit} by similarity desc";
+        // reorder to make it easier to ignore the embedding (key, metadata, timestamp, similarity, embedding
+        // Using tostring to make it easier to parse the result. There are probably better ways we should explore
+        similarityQuery += "| project " +
+            // Key
+            $"{s_collectionColumns[0].Name}, " +
+            // Metadata
+            $"{s_collectionColumns[1].Name}=tostring({s_collectionColumns[1].Name}), " +
+            // Timestamp
+            $"{s_collectionColumns[3].Name}, " +
+            "similarity, " +
+            // Embedding
+            $"{s_collectionColumns[2].Name}=tostring({s_collectionColumns[2].Name})";
+
+        if (!withEmbeddings)
+        {
+            similarityQuery += $" | project-away {s_collectionColumns[2].Name} ";
+        }
+
+        using var reader = await this._queryClient
+            .ExecuteQueryAsync(
+                this._database,
+                similarityQuery,
+                GetClientRequestProperties(),
+                cancellationToken
+            ).ConfigureAwait(false);
+        while (reader.Read())
+        {
+            var key = reader.GetString(0);
+            var metadata = reader.GetString(1);
+            DateTime? timestamp = reader.IsDBNull(2) ? null : reader.GetDateTime(2);
+            var similarity = reader.GetDouble(3);
+            var _embedding = withEmbeddings ? reader.GetString(4) : default;
```

Comment:
Same comment as above. Although it is possible with more effort to create a more "standard" way of doing things, I'm not sure the investment right now is worth it (nor am I sure it would make the code more "future proof")

Reply:
Sounds good, feel free to skip it.